### ISSUE TITLE
Add support for Leadshine R2EC coupler and associated digital IO modules, relay output modules, and encoder input modules

### ISSUE THREAD

**Issue by jrcdec** [linuxcnc-ethercat/linuxcnc-ethercat#434]
I am a total newbie on Github and haven't coded seriously for 30 years (did a little C back then), so please excuse any etiquette issues which will all be unintentional. I am in awe of and deeply respectful of all the amazing contributors and their work on this project.
Leadshine has a coupler (R2EC) much like the Beckhoff EK series, and a set of modules which can plug into it. 
The prices are very attractive and I (and I am sure many others) would like to be able to use them in machine projects. (I am currently working on 3 ). For example, the 16 input or output modules are about USD 36, and the 32 input or output ones are about USD67.
The digital I/O ones come in 16 and 32 input or output versions, with a 16 output relay output and a 16in/16 out as well. They also have dual encoder input modules and thermocouple and thermistor temperature sensing modules.
I have collected the xml files and manuals and made a crude attempt at copying some of the devices ID's etc into renamed copies of the Beckhoff lcec1xxx.c and lcec2xxx.c modules, which will need some tweaking but can give someone who (unlike me) knows what they are doing at least a bit of a head start.
From what I can see in the code, lcec.h will need an entry along the lines of #define LCEC_LEADSHINE_VID of 0x00004321  to set up the Leadshine vendor ID, and the R2EC will need entering probably into the lcec_passive.c along with its Beckhoff equivalents.
There may be a wrinkle in that the R2EC has a DeviceID but the modules have ModuleID's instead, and the SDO of the R2 then returns on Index 0xF050-00 the number of modules attached to the R2EC and then on 0xF0500-01, 02, 03 etc the Module ID of each connected module (the manual contains an error and says the Index is F030 but I have confirmed it's F050 using the ethercat upload command). I don't know if this is similar or not to how Beckhoff does things.
The PDOs then sequence themselves on 6000-61F0 at intervals of 0x10, according to the manual on page 14, for example. Don't know how this is handled when writing lcec files.
The encoder inputs will presumably need to go into lcec_class_enc.c somehow.
If anybody wants to have a go, I am more than willing to send him/her an R2EC and a 32-input module, a 32 output module, a 16 output relay module, and a 2 input encoder module for testing and as a bounty/encouragement if that is acceptable under the project rules.

Many thanks!
James
[R2 series Extension Module User Manual V1.0.pdf](https://github.com/user-attachments/files/15580109/R2.series.Extension.Module.User.Manual.V1.0.pdf)
[R2 series IO Module Catalogue.pdf](https://github.com/user-attachments/files/15580111/R2.series.IO.Module.Catalogue.pdf)
[lcec_pm00xx.txt](https://github.com/user-attachments/files/15580151/lcec_pm00xx.txt)
[lcec_pmxx00.txt](https://github.com/user-attachments/files/15580152/lcec_pmxx00.txt)
[R2EC-v1.4.0.0xml.txt](https://github.com/user-attachments/files/15580155/R2EC-v1.4.0.0xml.txt)

**Comment by pippin88** [comment]
They look like a good option.
New product I think?

What was the price for the R2EC coupler?

I can't help with actual coding sorry

**Comment by jrcdec** [comment]
Just to narrow things down a little, my main priorities are the digital input modules, the digital output modules (signal and relay) and the encoder input modules. 

The analog input modules and the thermocouple/thermistor input modules look very nice too but are of less immediate use, though I am sure they will also be useful to many people, and the digital mixed module (16 in/16 out) is also great but can be replicated with one 16 input module and one 16 output module, so it is less of a priority.

Both the digital input and output modules come in several models, but the only real variable is whether they are 16 or 32 inputs or outputs. The model variants beyond that only relate to whether they are NPN or PNP and in some cases to the connector type. The internal logic is identical so the driver can be exactly the same.

> They look like a good option. New product I think?
> 
> What was the price for the R2EC coupler?
> 
> I can't help with actual coding sorry

**Comment by jrcdec** [comment]
Re: pricing - the below are approx. FOB China prices -you would need to add shipping and any import or other local taxes wherever you are:
A few examples 👍 
R2EC coupler USD 59
PM-3200 (32 input) USD 67
PM-1600 (16 inputs) USD 36
PM-0032-N (32 output) USD 67
PM-0016-N (16 digital output) USD 36
PM-0016-R (16 relay outputs) USD 85
PM-E0200-D (dual encoder input (2 inputs), balanced lines input) USD 119 
PM-E0200-S (dual encoder input, single ended inputs) USD 108
PM-A0400-IV 4x analog input (16 bit A/D converter) USD 94
PM-A0004-IV  4x analog output (16 bit ADC) USD 101

**Comment by scottlaird** [comment]
Interesting.  I'll take a look in a bit, but I'm completely underwater with work at the moment, and that may not change much until late June.

**Comment by scottlaird** [comment]
Looking at these, they look interesting, and probably not *that* hard to support, but they'll end up having to all be part of the same driver, or we'd need to come up with a plugin mechanism for Leadshine R3 modules, which is more work than it's worth right now.  As mentioned above, it looks like the R2EC coupler is an EtherCAT device, but the individual modules aren't.  So we'd have a R2EC driver that grabbed it, probed what's connected, and registered pins for each of the sub-modules.  A year ago, that would have been painful.  Today, there's enough support for dynamic config that it's probably pretty easy, although there may not be an easy driver to copy up front.  The digital I/O ports at least look pretty trivial to implement.  I haven't looked at the encoder or analog modules, but they probably follow a similar-enough pattern.

Man, a $119 encoder input is *amazingly* cheap though -- the last time I checked used Beckhoff EL5xxxs were mostly $500+.

**Comment by jrcdec** [comment]
" Man, a $119 encoder input is amazingly cheap though -- the last time I checked used Beckhoff EL5xxxs were mostly $500+. "

My thoughts exactly. And it is a two-channel unit. The analog input and output ones are 4-channel units and use 16-bit converters, and all 4 channels can update at the 1KHz EtherCAT bus speed. And the temperature sensing ones (thermistor and thermocouple) are also all 4-channel units. It seems like the price point is exactly what we need for DIY and open-source projects.

On the digital input and output ones, Beckhoff are mostly 2,4, and 8 channel modules, with a few 16 channel. These Leadshines are either 16 or 32 channels, so one input and one output module will be enough to deal with all the digital I/O in all but the most absurdly complicated machines.

Late June works for me - I have received all my Leadshine kit already but am waiting on some of the mechanical components which had longer lead times, so I won't be able to start putting things together until early July anyway.

Judging by what you say about the module structure being slightly different it sounds like you are going to need to have some in your hands to test as you code. 

My offer stands - I would be delighted to send you a set of coupler, digital input, digital output, and encoder input to test with and to keep when you are done. Send me a shipping address privately and I can get it organized in the next week or two (more like 2 as I am going on a business trip next week).
I am working on putting together 3 machines at the moment (a mill, a 5 axis, and a lathe), and the savings on going for these relative to Beckhoff on 3 machines are so huge that the cost of sending you a set to test is tiny in comparison, and the rest of the community gets to benefit from the drivers. So to me it's a total no-brainer if you are interested.

If you want to go the whole hog and do the analog and the thermal ones too while you are at it, I can probably stretch to that too and include them in the shipment. My spindle has a thermistor output which I have not been using, and it would be great to be able to track it too, and there are a few projects in the back of my mind that could use the analog I/O and which I have been procrastinating on for years for lack of a good solution.

I am going to put together a text file with the results of probing my existing R2EC+module stack using the "ethercat" command and attach it to a message later today to show what I have found to date. The results are generally pretty logical but I did find a few quirks like the Index value in one of the manual entries being off from F030 to F050, and I can save you some time by giving you that data.

**Comment by jrcdec** [comment]
The attached file has the result of running a few probes into the R2EC + module chain using the "ethercat upload" command and a few others.
[r2test.txt](https://github.com/user-attachments/files/15602891/r2test.txt)

**Comment by jrcdec** [comment]
I also asked Leadshine about the F030 / F050 issue and got an email reply.
I will send you (Scott) the Leadshine email answering the query to the email address listed in your GitHub contributed files, but the gist of it seems to be that once the modules are "added" on a "master station" then they appear in both F030 and F050, but until that is done they only appear in F050. 
Omitting the pictures in the email the text says:

" Hi , James
Wish you have a nice day!
I tested the issue which you feedback in the previous Email, the manual is correct  but some important content was missed.
1.Regarding the SDO F030H, the description of these objects are correct, value of  F030H-00H is number of configured modules.
When you use the EtherCAT master station connect to the R2EC coupler and add some modules,  F030H-00H will be changed, refer to the below picture. "

(then there is a picture of some software screen) then

"2.Regarding the SDO F050H, it is the number of modules scanned, 
when you connect the module but haven't use the master station to add the modules to R2EC coupler, you can get the correct value through F050H, and value of F030H is 0.
Refer to the below pictures.
"
(more pictures)
"3.Generally, you can use the PLC programming software to configure the modules, or you can use the TwinCAT software as the master station to connect the modules.
"
I am insufficiently familiar with EtherCat master station or PLC software to understand that, but I am hoping that it makes more sense to others in the list. It appears to be an automatic result of connecting the coupler and modules to an ethercat master. Maybe once a linuxcnc-ethercat driver registers the R2EC all this happens automatically?

**Comment by stean111** [comment]
Are your modules working already ? 

I`m using similar IO modules successfully from Veichi company.

Looking at the xml files I think they are the same. These companies either use a common platform or copy each other's (also Inovance etc...).
The modules work correctly but I used TWINCAT help to preparing XML file, and initially after a lot of testing they work fine.

I`ll send my files to help you.

**Comment by jrcdec** [comment]
That would be great, please send the files. I have only managed to get as far as is shown in the messages above. Do the modules you mention also have a similar coupler arrangement to the R2EC? If they do it would be great to have a solution for everyone that allows them to be used with LinuxCNC and provide easy access to all the pins.

**Comment by jrcdec** [comment]
I replied directly on GitHub but don’t know if you received it, and I have been out of commission for a few days due to a nasty hard disk crash requiring a total rebuild of my main PC.

I have only managed to get as far as the posts on GitHub with the R2EC coupler and IO modules, so any help is more than welcome – would love to receive your XML files and any simple working example to get me started.

I had a look online at the Veichi modules you mention and they look a little similar, though they are mainly 8 and 16 channel ones as opposed to 16 and 32 channels on the Leadshine ones. However, if the underlying chipset is similar they may well be cross-compatible. I wonder if the coupler works the same way too.

 

If Scott’s schedule is getting a bit freer as we approach the end of June I would love to see what we can do to get them working in LinuxCNC.

**Comment by jimwhiting** [comment]
I'm also interested in getting the R2EC stuff working under LinuxCNC, do you have a link for where the modules can be purchased? I'm not having much luck with google!

**Comment by jrcdec** [comment]
I bought them directly from their sales email – ***@***.*** ***@***.***>    . They assigned a salesman ***@***.*** ***@***.***> ) and I put a list together, he quoted me prices, and I then paid them by bank transfer. 

As I live in Hong Kong there were no duties since Shenzhen is just across the mainland China border and only a few Km away. 

Depending on where you are based there may be some import taxes and paperwork to deal with. 

Their quotes are FOB Shenzhen, China, so you need to pay for shipping separately or give them a Fedex or equivalent account to put it under. I suppose they might also be able to quote with freight if you ask them in advance.

**Comment by jimwhiting** [comment]
Thanks, my local Leadshine disty has just let me know they can get them. They have also introduced me to an alternative line from Kinco as well.

**Comment by jrcdec** [comment]
The addresses above got obscured by the forum software filter, but it was basically “sales” at the company’s name dot com, and the other one was “john”, same deal. I have found john to be quite helpful. He got some of his technical people to send me some xml files which I could not find on the normal downloads from their website

**Comment by stean111** [comment]
sorry, I`m on the road now, so I`ll send today or tommorow, I`m sure they have the same copy/paste things because by mistake company sent me an xml file from another company for their modules, oops ...

**Comment by stean111** [comment]
Hello, enclosed I send my ini files. IO`s are second slave 2 x 16 inputs and 2 x 16 outputs (first slave are servodrive). Importants are xml inits commands.
[linuxcnc-cia402-main.zip](https://github.com/user-attachments/files/15936194/linuxcnc-cia402-main.zip)

**Comment by stean111** [comment]
yes Veichi producing also analog modules but not encoder modules yet, so its good news if Leadshine offer also encoders inputs, its also interesting for me, I`ll buy and check it out.

**Comment by stean111** [comment]
James, I can`t find encoders modules on Leadshine website, please help me find it.

**Comment by jrcdec** [comment]
I just had a look and you are right, the encoder modules don’t appear as individual items in the Products>Controls>R2Series section of the website, which only has the coupler and the digital and analogue IO modules. But if you go to Products>Controls>R2Series and go to the Downloads section, then Catalogs in the Downloads section, you get a catalog of the whole range which does list the encoders and the temperature modules, both of which are omitted from the individual product listings. They have two encoder modules, They are both dual-channel units (can accept two encoder inputs), and the only difference is that one takes single-ended inputs and the other one takes balanced inputs and so has better noise immunity. The price difference is small so it’s worth going for the balanced line inputs if your encoder supports it.

**Comment by stean111** [comment]
I see, thanks. In order to build xml init commands for your modules you need to install a free Twincat v2 software (not v3).

**Comment by jrcdec** [comment]
Many thanks

I had a quick look and if I am not mistaken you have implemented the Veichi modules in cia402.hal as the “wejsciowy” input and “wyjsciowy” output entries – am I right?

 

You seem quite a lot more up to speed than I am so I would definitely welcome some help. 

When you say “In order to build xml init commands for your modules you need to install a free Twincat v2 software (not v3)”, which specific one do I need? The x64 Engineering package? I have downloaded that one from the Beckhoff website downloads section but it seems to be a 30-day evaluation version rather than a free licence. 

 

And once I install it, how do I use it to generate the right xml init commands to use in LinuxCNC?

Please assume I am quite clueless on TwinCAT (which is the truth) and give me a basic “for dummies” version of the required steps if possible. Many thanks!

**Comment by stean111** [comment]
So its from the Polish language, “wejsciowy” input and “wyjsciowy” output, you`re right.

I`m using Twincat 2.11 from Beckhoff website. Its 30-day evaluation but interesting functionality works after the trial period too.

Please send me your xml file I`ll show you.

I checked Leadshine website so R3 version is exactly same like a Veichi 12 mm slim modules.

**Comment by stean111** [comment]
Enclosed I send, XML file generated by TwinCAT for one module PM-1600.  

[Box 1 (R2EC).xml.txt](https://github.com/user-attachments/files/15971889/Box.1.R2EC.xml.txt)

**Comment by jrcdec** [comment]
This looks extremely promising!

How do you do it?

I don’t have a PM-1600 t test with, so I can’t check if this one works, but I do have many other Leadshine modules

Could you do the same and send individual file for the following modules so I can try to test the files? (all modules are in the big Leadshine R2EC-v1.4.0.0.xml file in their website)

 

R2EC

PM-3200

PM-0032

PM-E0200-D

PM-0016-R

 

I can test all of the above, and it would also be nice to get files for the following:

 

PM-A0400IV

PM-A0004-IV

PM-T0400-TC

PM-1616

 

Many thanks!

**Comment by stean111** [comment]
yes, one xml file for chain in the correcr order.

lets try one module first, so I generating now for one PM-3200.

[Box 1 (R2EC).xml.txt](https://github.com/user-attachments/files
[startup.csv](https://github.com/user-attachments/files/16039859/startup.csv)
/16039854/Box.1.R2EC.xml.txt)

You need to change these files appropriately for Linux to process them. I can help do this maybe remotely, I have a big problem with not enough time.

**Comment by stean111** [comment]
we can try it if you have a time today, I don`t know what is your local time, now my local time is UTC+02:00

**Comment by jrcdec** [comment]
Many thanks!  I am in Hong Kong so my time zone is some hours ahead of Central Europe. I am on a business trip this coming week but will be around on the week of the 8th to 13th and able to experiment.
I have downloaded TwinCAT v2.11 Engineering and would be very interested in knowing how you generate these files with it before the licence free trial runs out.
I imagine you somehow need to open the LEadshine R2 series xml file and then somehow tell TwinCAT which modules you have and want to keep for an exported file?

**Comment by scottlaird** [comment]
Okay, FWIW, I have a set of modules on hand now (thanks @jrcdec!), and I can detect them, but the PDO config is currently confusing me a bit.  It says that the PDO mapping can't be configured, but the ESI XML file kind of implies that it happens automatically if we set it up right.   Presumably this is the same process that copies configs from 0xf050 into 0xf030?

**Comment by jrcdec** [comment]
I know a lot less than everyone else here, so I could be proposing something entirely nonsensical or hopelessly hard, but is it possible that if you do a driver to initialize the R2EC coupler through the Init. / Pre-Op / Safe-Op/ Operational sequence and set it to the Operational state, that triggers the process which populates the 0xf030 PDO?

**Comment by jrcdec** [comment]
Alternately, does the above message from stean111 provide a valid initalization mechanism? He seems to have had success with what he thinks is a related family of couplers.

**Comment by jrcdec** [comment]
stean111  -I am back at my desk and can work on this today or probably better tomorrow if you have time. I have installed Beckhoff TwinCAT 2.11 but don't know how to use it, down to the basics of for example how to tell it which Ethernet port to use to connect to the modules. Once I get through that I would love to learn how you generated the sub-files based on the big "all modules" Leadshine R2 xml file, and how to use the resulting sub-files on LinuxCNC.

**Comment by scottlaird** [comment]
I made another pass through the assorted XML files, and I think I'm still missing something.  If I set up PDOs and 0xF030 as shown while in state=PREOP, then  the config "sticks" -- I'm able to view the SDOs and see the values that I wrote.

But, transitioning to state=SAFEOP resets 0xf030:* back to 0.  And once it's in SAFEOP the SDOs aren't writeable.

Here's a shell script version of what I'm trying to do; this is basically the same as the info that @stean111 provided above.  It only configures 1 module, while I have 4.  Extending the same thing to 4 modules (with the correct IDs) behaves identically.

```
#!/bin/bash

SET="ethercat -m 1 download -p 6"

$SET 0x1c12 0 0
$SET 0x1c13 0 0
$SET 0x1c13 1 0x1bda
$SET 0x1c13 2 0x1a00
$SET 0x1c13 0 2
$SET 0xf030 0 0
$SET 0xf030 1 0x61100045
$SET 0xf030 0 1
```

It feels like I'm still missing a step here somewhere.

**Comment by jrcdec** [comment]
I have sent the one Leadshine application engineer I have contacted previously an email which reproduces Scott’s message above (about settings not “sticking” when transitioning to SAFEOP) almost verbatim to see if they have any suggestions. Will post their reply as soon as I hear back. Not holding my breath since the application engineer and his team seem to mainly use software tools like TwinCAT rather than actual low-level register writes, but you never know. Someone must be doing the low-level coding so we might get lucky and reach them.

**Comment by jrcdec** [comment]
@scottlaird and everyone - as promised I wrote to Leadshine and they have already replied with some questions and clarifications.
My message to them was basically Scott's message above plus the following questions about their new R3 series which was released midway through this thread's existence:
"I notice you recently released an R3 series. Does the initialization and addressing logic work the same way as the R2 series? Are the prices similar? And, is the R2 series going to be discontinued or will both series continue to be offered indefinitely?"

Leadshine's reply was:
"1 Regarding the question, can you provide some picture or video to reference?
2. R3 series module is similar with R2 series, you can check the attached manual. R3 series is ultra-thin type module and the prices are slightly cheaper than R2 series.
3. Both series module will continue to be offered, if you have new project, we recommend using the R3 series module.
"
[R3 series Extension Module User Manual_V1.00.pdf](https://github.com/user-attachments/files/16175567/R3.series.Extension.Module.User.Manual_V1.00.pdf)

In summary : 
-	The engineer wants some pictures or a video showing what is described in Scott's message above. I suspect they are assuming TwinCAT or similar rather than Linux command line, but no harm in showing them what happens at each step to see what they propose trying. 
-	They have a new R3 series which was launched literally as this thread was getting going
-	The new R3 series seems to be very similar, so they work done on R2 should carry over
-	The new R3 series is both thinner AND claimed to be “slightly cheaper”
-	They prefer it for new designs, though both series will continue to be offered
-	However, R3 can’t currently be seen as a comparable replacement for R2 because:
-	the new series’s I/O modules are 8 and 16 channel units, rather than 16 and 32 channels in the R2 series.
-	And also the new series doesn’t seem to have relay output units or encoder input units, 
-	One improvement is that the new series introduces a new serial communications output unit (ModBusRTU etc), but this hardly seems to make up for the lack of encoders and relays, so I suspect the R2 series is here to stay.

I am attaching the R3 manual for anyone who is interested.

**Comment by jrcdec** [comment]
@stean111 - many thanks for your kind offer again. I am free this Saturday morning Asia time, but after that I am travelling until the 10th of August, so It might be best to wait until then. And with any luck Scott will have made some progress too in the meantime.

**Comment by jrcdec** [comment]
@scottlaird and everyone. I have managed to get the R2 and two modules to reach the "RUN" mode. What I did is ask Leadshine for a simple sample program for their MC516 PLC (of which I had ordered one to play around with) which would connect by an EtherCAT port to an R2EC daisy chained with an RM0032 output module and an RM3200 input module.

They sent me a sample program and after a LOT of configuring of their Leadsys Studio software, installing the right libraries, fussing with connection settings etc., I finally managed to connect today and compile and load the program. The R2EC coupler now shows 3 green lights - POWER, SYS, and RUN, and the ERR light is not lit, which is quite promising.

My thinking is that now that I can get the R2EC to reach the RUN state, it should be possible to snoop into the actual EtherCAT packets by connecting the PLC to a switch and the switch to the R2EC coupler and using a Linux packet logging/inspection utility plugged into the same switch to record the messages between the two as the R2EC gets activated by the PLC. Those packets should then be able to tell the cleverer ones amongst us (probably Scott, certainly not me!) what EtherCAT messages perform the initialization magic, so that they can be reproduced in the driver.

Does that sound like a possible way forward, and if so, what package should I try by way of Linux packet logger/inspector and with what settings to record the initialization traffic?

Thanks!

**Comment by scottlaird** [comment]
Okay, first, to collect traffic the easiest way is to use something like `tcpdump -i $INTERFACE -s0 -w $FILE` to record all traffic on `$INTERFACE` (replace with `eth0` or whatever your machinee uses) to `$FILE` (replace with a real file name).

The problem is if you have a switch in the middle, then Linux isn't going to see much, because that's how switches work.  Although EtherCAT is weird enough that maybe it will anyway.  EtherCAT plays games with MAC addresses.

The simplest way to deal with this might be to replace the switch with a Linux PC with 2 Ethernet interfaces, and then put the two interfaces into "bridge" mode, so that it acts like a 2-port switch.  Then it (obviously) could see everything.  That's getting a bit complex, though.

Is the PLC program shareable?  

Alternatively, once the R2EC is in a working state, it *might* be good enough to dump all of the random SDOs from Linux.  There's an (undocumented) script in `scripts/sdos.sh` that might be good enough.

**Comment by jrcdec** [comment]
On 3 Sep 2024, at 21:44, Scott Laird ***@***.***> wrote:﻿
Okay, first, to collect traffic the easiest way is to use something like tcpdump -i $INTERFACE -s0 -w $FILE to record all traffic on $INTERFACE (replace with eth0 or whatever your machinee uses) to $FILE (replace with a real file name).
The problem is if you have a switch in the middle, then Linux isn't going to see much, because that's how switches work.  Although EtherCAT is weird enough that maybe it will anyway.  EtherCAT plays games with MAC addresses.
The simplest way to deal with this might be to replace the switch with a Linux PC with 2 Ethernet interfaces, and then put the two interfaces into "bridge" mode, so that it acts like a 2-port switch.  Then it (obviously) could see everything.  That's getting a bit complex, though.
Is the PLC program shareable?
Alternatively, once the R2EC is in a working state, it might be good enough to dump all of the random SDOs from Linux.  There's an (undocumented) script in scripts/sdos.sh that might be good enough.

**Comment by jrcdec** [comment]
I have obtained screenshots of the logs in LeadSys Studio showing some possible messages sent to the R2EC by the Leadshine 
 Master in the PLC when the R2EC was being initialized. I also did as @scottlaird suggested and used tcpdump on a Linux box connected in the same switch that was used between the PLC's EtherCAT output and the R2EC's EtherCAT input as the R2EC was being initialized. Over what was probably a few seconds it captured around 2832 packets until I stopped it with Control-C. :
jrc@ot08:~$ sudo tcpdump -i enp0s31f6 -s0 -w packetlog
tcpdump: listening on enp0s31f6, link-type EN10MB (Ethernet), snapshot length 262144 bytes
^C2796 packets captured
2832 packets received by filter
0 packets dropped by kernel
 It seems like it did manage to capture stuff despite the switch. I also did an experiment and there were no packets captured before I asked LeadSys Studio to tell the PLC to get started, so we may have got lucky and not been blocked by the switch.
I am attaching the packet log file too. It originally had a .dat extension and I had to change it to .txt to get the forum software to upload it
![Screenshot 2024-09-04 111149](https://github.com/user-attachments/assets/c85ec8da-85b4-4089-815a-03234c732916)
![Screenshot 2024-09-04 112534](https://github.com/user-attachments/assets/ae235b88-5398-40ba-8c54-2b025a02b228)
[packetlog.txt](https://github.com/user-attachments/files/16860711/packetlog.txt)

**Comment by jrcdec** [comment]
One more thing - I had a look for the sdos.sh script in /home/cnc/linuxcnc-dev/scripts and could not find sdos.sh - an alphabetical listing went straight from runtests.in to setup_designer.in
Is there a different, more up-to-date linuxcnc-dev that has this script (and presumably other stuff as well)?
Thanks!

**Comment by jrcdec** [comment]
One more brief thing. I opened the packet log file in Wireshark and it recognized the packet format of all the packets as being ECAT. One more clue suggesting that it may have worked despite the intervening switch…

**Comment by scottlaird** [comment]
sdos.sh: https://github.com/linuxcnc-ethercat/linuxcnc-ethercat/blob/master/scripts/sdos.sh

I'll look at the packet dump and see if anything jumps out, but I probably can't do much today.

**Comment by jrcdec** [comment]
Question for @stean111 - what do all the SDO writes to 1C12 and 1C13 do in your startup script? If I look at the screenshots I got above using Leadshine LeadSys software to connect to the modules using a Leadshine PLC (and which successfully initialized the modules), the logs also show writes to 1C12 and 1C13. In particular, 1BFA gets written to 1C13 subindex 1 (same as in your startup.csv file) and 1A10 gets written to 1c13 subindex 2 (very slightly different from your file, where 1A00 gets written instead of 1A10)

**Comment by stean111** [comment]
Sorry for late answer... 

> what do all the SDO writes to 1C12 and 1C13 do in your startup script?

its here in file st_I16x2_O16x2.xml :

<?xml version="1.0" encoding="utf-8"?>
<EtherCATMailbox>
	<CoE>
		<InitCmds>
			<InitCmd Fixed="1" CompleteAccess="1">
				<Transition>PS</Transition>
				<Comment><![CDATA[SubIndex 000]]></Comment>
				<Timeout>0</Timeout>
				<Ccs>1</Ccs>
				<Index>7186</Index>
				<SubIndex>0</SubIndex>
				<Data>0300A01710161816</Data>
			</InitCmd>
			<InitCmd Fixed="1" CompleteAccess="1">
				<Transition>PS</Transition>
				<Comment><![CDATA[SubIndex 000]]></Comment>
				<Timeout>0</Timeout>
				<Ccs>1</Ccs>
				<Index>7187</Index>
				<SubIndex>0</SubIndex>
				<Data>0300A01B001A081A</Data>
			</InitCmd>
.....


index 7187 its dec so means in hex 1C13 ...

please check all my cofiguration files in main .xml ethercat2-conf_I16x2_O16x2.xml there is line to process int command :
...
<slave idx="1" type="generic" vid="00850104" pid="0x29800000" configPdos="true">
      <initCmds filename="st_I16x2_O16x2.xml"/>
...

so I did it but not fully automatically I did it by hand but TwinCAT help me to do it.

**Comment by stean111** [comment]
Twincat producing lists like this its step by step what we need to configure modules:

`Transition;Protocol;Index;Data;Comment
<PS>;CoE;0x1C12 C 0;03 00 A0 17 10 16 18 16;download pdo 0x1C12 index
<PS>;CoE;0x1C13 C 0;03 00 A0 1B 00 1A 08 1A;download pdo 0x1C13 index
<PS>;CoE;0xF030 C 0;04 00 05 00 21 20 05 00 21 20 08 00 21 20 08 00 21 20;download slot cfg
PS;CoE;0xF800:01;0x00 (0);Configuration error run state
PS;CoE;0xF800:02;0x00 (0);Slave and module error run state
PS;CoE;0xF800:03;0x01 (1);EtherCAT commucation error mode
PS;CoE;0xF800:04;0x00 (0);ESC error counter behavor
PS;CoE;0x8000:01;0x04 (4);DI module CH0 filter time(0.25ms)
PS;CoE;0x8000:02;0x04 (4);DI module CH1 filter time(0.25ms)
PS;CoE;0x8040:01;0x04 (4);DI module CH0 filter time(0.25ms)
PS;CoE;0x8040:02;0x04 (4);DI module CH1 filter time(0.25ms)
PS;CoE;0x8080:01;0xFF (255);DO module CH0 stopmode after EtherCAT lost link
PS;CoE;0x8080:02;0xFF (255);DO module CH1 stopmode after EtherCAT lost link
PS;CoE;0x8081:01;0x00 (0);DO module CH0 stopvalue after EtherCAT lost link
PS;CoE;0x8081:02;0x00 (0);DO module CH1 stopvalue after EtherCAT lost link
PS;CoE;0x80C0:01;0xFF (255);DO module CH0 stopmode after EtherCAT lost link
PS;CoE;0x80C0:02;0xFF (255);DO module CH1 stopmode after EtherCAT lost link
PS;CoE;0x80C1:01;0x00 (0);DO module CH0 stopvalue after EtherCAT lost link
PS;CoE;0x80C1:02;0x00 (0);DO module CH1 stopvalue after EtherCAT lost link`

**Comment by jrcdec** [comment]
@stean111 , many thanks for persevering! 
I am travelling this week but if you let me know how to message you so I can get a shipping address etc, next week when I get back I can send you a set of modules like I did with Scott so you can try it out and hopefully we get a final working result. I have  been in touch with Leadshine also with the same screeenshots I posted here but their answers were a lot more general than your detailed work above, so I am quite hopeful that you can solve it at the xml leve for linuxcnc and then Scott can use the data to get some working drivers done for all of us. .

**Comment by jrcdec** [comment]
@stean111 - I am back from my trip and can send you some modules to test if you let me know how to reach you.
@scottlaird and @stean111 - I have received a video and some screenshots from Leadshine showing what shows up in the various PDOs in Codesys, and I am attaching it in case it's useful - I am not familiar with it but it does seem to pop out a lof of detailed internal settings that may be key to getting the driver working - hopefully they can be the key to getting it all working? 

https://github.com/user-attachments/assets/ea2642dd-5320-42fc-8437-51517ef09dd0

![PDO_INPUT](https://github.com/user-attachments/assets/b8157cba-afe1-4a0a-928d-60a9dfca9933)
![PDO_OUTPUT](https://github.com/user-attachments/assets/63bb8d03-ffb2-4117-90b2-e06944a5002e)

**Comment by stean111** [comment]
> I am back from my trip and can send you some modules to test if you let me know how to reach you.

I thinking how to do it, we needs to find a cheap courier or by regular post.
I`ll send you my address (in private message if it is possible here), for me interesting is encoder module, becasue Veichi not producing it yet , of course I`ll run for you what you needs

**Comment by jrcdec** [comment]
@stean111  - No private messages on GitHub - if you make your email non-private in your GitHub profile I can see it here. Or you can send it to me - the first 3 letters of my username here at a domain that begins with aaam then a full stop then the initials of Hong Kong

**Comment by stean111** [comment]
ok I sent contact email

**Comment by stean111** [comment]
Hello. I want to return to the topic. I apologize for not responding for so long time.
During the tests on Leadshine (thanks to jrcdec) and Veichi, a new problem appeared, which I will describe.

IO systems, such as WAGO, Leadshine, Veichi, have a modular system, which means we have one Slave to which we can connect many IO modules.
This requires configuration in transit mode from PREOP to OP. We need to load which modules are connected, assign them addresses, and configure individual modules.

I manually created the appropriate commands, which I attached for review. 

Everything works "almost" well.

The problem is that during the first start-up (despite the initialization commands working correctly), errors appear.
These errors mean that we cannot configure specific modules because during the transit, they are not yet ready for configuration. The Slave does not see the modules yet, so it cannot configure them, and when we cannot configure the input module, the module reports OP status and seems to work, but it does not see when we input logical states.

It's like the configuration is happening too fast.

However, when I restart LinuxCNC again, the modules are visible to the slave, and the configuration is saved correctly, and everything works properly.

Instead of restarting LinuxCNC, I can also enter the configuration commands manually in the terminal, and it will also work.

Soon, I will attach the configuration for Veichi as an example, as it works the same as with Leadshine. Jrcdec I`ll also send you the configuration for Leadshine, but unfortunately, they will only work after restarting LinuxCNC (but working like this).

**Comment by jrcdec** [comment]
Many thanks for your persistence! I look forward to your results.
One question - if re-entering the commands manually on the terminal works, is it possible to do that in code in the driver and effectively issue them once, pause, and reissue the same commands again, or something like that?
Alternately, you say it may be happening "too fast". Do things improve if you add short pauses in the code between successive initialization commands? Maybe a silly loop that does nothing except count up and down to 10000 between commands, for example?
Or, since the packet cycle time on EtherCat is usually once per millisecond, maybe make sure that the successive initialization commands are spaced at least a few milliseconds apart, or something like that.? It could be something as silly as an internal time constant/initialization time inside the coupler hardware which gags and errors-out when it's fed initialization data too quickly before it has time to execute each step completely, including reading and processing initialization success and status messages from the individual modules etc.. 
For example, if its internal logic is such that it can only initialize one module per EtherCAT packet and each packet comes every 1 msec, then the module's response takes another 1msec (or more) cycle, etc., trying to initialize more than 1 at once may cause a bottleneck.

**Comment by AhmadEmaraPrecisionTeq** [comment]
hello guys,
i bough the new R3EC + DI + D0 + Analog output modules, i faced the same problem as you guys, is the problem solved, regarding the F030 problem what i discovered from the manual that it should be configured to match the F050, so what i did is i ran some terminal commands to check the values of F050 index 0,1,2,3 and used command "sdoconfig" in my XML file to set f030, without any PDO mapping and when i ran the configuration it opens normally and the ER led goes off in the modules, now i am stuck in the PDO mapping part, can anyone here helps me ?

**Comment by grandixximo** [comment]
If I'm understanding correctly: the R2EC coupler exposes connected modules via SDO 0xF050 (scanned count). For modules to work, 0xF030 (configured count) has to be written to match, plus a few per-module setup writes. stean111 got this going manually in April with init commands; AhmadEmaraPrecisionTeq replicated the F050/F030 step in December but hit the same PDO mapping wall. Two reasonable paths: document the sdoConfig recipe that works as a copy paste template, or write a proper R2EC coupler driver that does F050 read, F030 write, and per module setup automatically. @stean111 do you still have your working XML / init commands and can you share the latest version, that's the seed for both options. @AhmadEmaraPrecisionTeq @jrcdec is anyone here able to test a draft coupler driver on real hardware? I can write one but need someone with the hardware to verify.

**Comment by jrcdec** [comment]
Many thanks for persevering! I have a set of coupler and modules, and scottlaird also has a set (I sent it to him), so there are people here who can try out your draft.

**Comment by grandixximo** [comment]
drafted #479
please test

**Comment by jrcdec** [comment]
That was quick!
What do I need to do to test it, and how do I remove it after testing?

**Comment by grandixximo** [comment]
@jrcdec testing steps + how to remove are in #479: https://github.com/linuxcnc-ethercat/linuxcnc-ethercat/pull/479#issuecomment-4370485335

**Comment by SoloReverse** [comment]
Has anyone tested this? I can't figure out what is wrong. perhaps i have a broken module. 
I have an R3EC + R3-0808-N-V20 (according to the xml) and i rewrote #479 to work with the R3EC, the outputs on the R3-0808-N work but the inputs are not registering. Anyone has any ideas, i'd love to hear them.

**Comment by SoloReverse** [comment]
After pulling what few hairs are left on my head. i got the driver working with the module. 
on top of mirroring `0xF050` into `0xF030`
i just have to make it work for any module now. i'll focus on the R3EC, but anyones welcome to help with the R2EC.
I added: 
**_what is bold is what i hard coded for my module to work_**

1. Clearing SDO `0x1C12:00` - SM2 Output PDO assignments = **_0_**
2. Clearing SDO `0x1C13:00` - SM3 Input PDO assignments = **_0_**
3. Setting SDO `0x1C12:N` with the value of the output mapping mode = **_0x1600_** 
4. Setting SDO `0x1C13:N` with the value of the output mapping mode = **_0x1A00_**
5. Setting SDO `0x1C12:00` with the number of `0x1C12:N` mapped above = **_1_** 
6. Setting SDO `0x1C13:00` with the number of `0x1C12:N` mapped above = **_1_**

current things to add. 

- [ ] SM2 and SM3 output mapping mode assignments shift by `0x08` per slot (as opposed to the actual data which shifts `0x10` per slot, however for the R2EC all shifts are `0x10`)
- [ ] most modules i have checked in the ESI output mappings are `0x1600` and `0x1A00`. I will make sure and might possible make that the default for now. 
- [ ] More testing on R2EC and R3EC inputs and outputs. 
- [ ] analog, temperature, encoders, and com are yet to be written. 

I have ordered all the modules to try and test them all to work. I would love it if I could get the RS485 Modbus to work.
the driver will end up dense however based on how slots map and not like Beckhoff modules where each is an EtherCAT slave, might need to be split up into multiple files?

@grandixximo I'd like to create a draft PR on top #479 for people to test it.

Here also is the latest xml grabbed from the chinese website. 
[R3EC-v2.4.0.0.xml](https://github.com/user-attachments/files/29353581/R3EC-v2.4.0.0.xml)

**Comment by grandixximo** [comment]
Go ahead, I don't think anyone tested it before, so thank you

**Comment by SoloReverse** [comment]
I have submitted a new draft PR #491 . Would be nice to have people test it out.
